File: problems/problems_3599/solution.py
from functools import cache

from math import inf

import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.minXor(*test_input)

    def minXor(self, nums: List[int], k: int) -> int:
        n = len(nums)
        pre_xor = [0] * (n + 1)
        for _i in range(n):
            pre_xor[_i + 1] = pre_xor[_i] ^ nums[_i]

        @cache
        def dfs(i, k):
            if k == 1:
                return pre_xor[n] ^ pre_xor[i]
            ans = inf
            for j in range(i+1, n - k + 2):
                ans = min(ans, max(pre_xor[i] ^ pre_xor[j], dfs(j, k - 1)))
            return ans

        return dfs(0, k)

File: problems/problems_3599/solution.go
package problem3599

import (
	"encoding/json"
	"log"
	"math"
	"strings"
)

func minXor(nums []int, k int) int {
	n := len(nums)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, k+1)
		for j := range dp[i] {
			dp[i][j] = math.MaxInt32
		}
	}
	preXor := make([]int, n+1)
	for i := range n {
		preXor[i+1] = preXor[i] ^ nums[i]
	}
	dp[0][0] = 0
	for i := range n {
		dp[i+1][1] = preXor[i+1]
	}
	for i := range n {
		for j := 2; j <= k; j++ {
			for l := j - 1; l <= i; l++ {
				xorValue := preXor[i+1] ^ preXor[l]
				dp[i+1][j] = min(dp[i+1][j], max(dp[l][j-1], xorValue))
			}
		}
	}
	return dp[n][k]
}

func Solve(inputJsonValues string) any {
	inputValues := strings.Split(inputJsonValues, "\n")
	var nums []int
	var k int

	if err := json.Unmarshal([]byte(inputValues[0]), &nums); err != nil {
		log.Fatal(err)
	}
	if err := json.Unmarshal([]byte(inputValues[1]), &k); err != nil {
		log.Fatal(err)
	}

	return minXor(nums, k)
}
